### PR TITLE
Implement `FlaxModelMixin`

### PR DESCRIPTION
**Pull request by mishig25**
### Implemented `FlaxModelMixin`

[FlaxModelMixin](https://github.com/huggingface/diffusers/blob/95073e13b2172bcadddd4ed528360c4f216d753b/src/diffusers/modeling_flax_utils.py#L39) is flax equivalent of [ModelMixin](https://github.com/huggingface/diffusers/blob/4e2c1f3a4da7436a861dd444035e6e27a3f1c6b7/src/diffusers/modeling_utils.py#L115) (i.e. a mixin that has `save_pretrained`, `from_pretrained`, & various checkpoints util functions).

After being discussed at https://github.com/google/flax/discussions/2454#discussioncomment-3627986, design-wise, [diffusers.FlaxModelMixin](https://github.com/huggingface/diffusers/blob/95073e13b2172bcadddd4ed528360c4f216d753b/src/diffusers/modeling_flax_utils.py#L39) differs from [transformers.FlaxPreTrainedModel](https://github.com/huggingface/transformers/blob/bce36ee065f7749c997be4c30a3f9279df7d5dba/src/transformers/modeling_flax_utils.py#L157) in the fact that: model class should be "state-less" https://github.com/google/flax/discussions/2454. 

see this comment for testing: https://github.com/huggingface/diffusers/pull/493#issuecomment-1246664508

### Example usages:

Class definition (find example [here](https://github.com/mishig25/stable-diffusion-jax/blob/d1f7b8f4f05bec98ebb2179d2b515e5caf73f33d/stable_diffusion_jax/modeling_unet2d.py#L572-L584))

```py
@flax_register_to_config
class UNet2D(nn.Module, FlaxModelMixin, ConfigMixin):
    sample_size:int=32
    in_channels:int=4
    out_channels:int=4
    down_block_types:Tuple=("CrossAttnDownBlock2D", "CrossAttnDownBlock2D", "CrossAttnDownBlock2D", "DownBlock2D")
    ....
    dtype: jnp.dtype = jnp.float32

    def setup(self):
           .....
```

from_pretrained (find example [here](https://github.com/mishig25/stable-diffusion-jax/blob/30ee5b06e7d12ba07bc6decfd9978626eab56eb2/example.py#L33))

```py
unet, unet_params = UNet2D.from_pretrained(f"{flax_path}/unet", dtype=dtype)
```

call `apply` just like any other `flax.linen.Module` (example [here](https://github.com/mishig25/stable-diffusion-jax/blob/30ee5b06e7d12ba07bc6decfd9978626eab56eb2/stable_diffusion_jax/pipeline_stable_diffusion.py#L70-L73))
```py
# predict the noise residual
noise_pred = self.unet.apply(
    {"params": inference_state.unet_params}, jnp.array(latents_input), jnp.array(timestep, dtype=jnp.int32), encoder_hidden_states=context, rngs={}
)
```

### Une Question

In [FlaxModelMixin.from_pretrained](https://github.com/huggingface/diffusers/blob/95073e13b2172bcadddd4ed528360c4f216d753b/src/diffusers/modeling_flax_utils.py#L195), I have NOT added section about checking [missing_keys & mismatched_keys like in trasnformers.flax_from_pretrained](https://github.com/huggingface/transformers/blob/bce36ee065f7749c997be4c30a3f9279df7d5dba/src/transformers/modeling_flax_utils.py#L828-L851). 

In short, _checking missing_keys & mismatched_keys_  functionality works by:
1. `random_state = ModelCls.init_weights()`
2. `pretrained_state = from_pretrained()`
3. compare keys between `random_state` VS `pretrained_state`

Therefore, to add it, it will require:
```py
# add new `input_shape` property to model cls
@flax_register_to_config
class UNet2D(nn.Module, FlaxModelMixin, ConfigMixin):
    input_shape: Tuple = (1, 32, 32, 4),

# require implementation of init_weights
class ModelMixin():
    def init_weights(self, rng: jax.random.PRNGKey, input_shape: Tuple, params: FrozenDict = None) -> Dict:
        raise NotImplementedError(f"init method has to be implemented for {self}")
```

and probably call `init_weights` inside `from_pretrained` since we do NOT have `FlaxModelMixin__init__`

Please let me know if I should add this functionality. wdyt

### TODOS
- [x] all docstrings need to be changed
- [x] soft flax dependency check?

cc: @kashif somehow I can't add you as a reviewer

### CODE REVIEW

**Comment by pcuenca** [comment]
I love this. I'll test it with my changes in #485.

**Comment by mishig25** [comment]
@pcuenca please let me know if you have any questions. `Example usages:` section in the description should contain all the necessary info.Moreover, you can just run example.py in this branch if you need: https://github.com/patil-suraj/stable-diffusion-jax/pull/10

**Comment by mishig25** [comment]
In [FlaxModelMixin.from_pretrained](https://github.com/huggingface/diffusers/blob/95073e13b2172bcadddd4ed528360c4f216d753b/src/diffusers/modeling_flax_utils.py#L195), I have NOT added section about checking [missing_keys & mismatched_keys like in trasnformers.flax_from_pretrained](https://github.com/huggingface/transformers/blob/bce36ee065f7749c997be4c30a3f9279df7d5dba/src/transformers/modeling_flax_utils.py#L828-L851). 

Please let me know if I should add this functionality. To add it, it will require:
```py
# add new `input_shape` property to model cls
@flax_register_to_config
class UNet2D(nn.Module, FlaxModelMixin, ConfigMixin):
    input_shape: Tuple = (1, 32, 32, 4),

# require implementation of init_weights
class ModelMixin():
    def init_weights(self, rng: jax.random.PRNGKey, input_shape: Tuple, params: FrozenDict = None) -> Dict:
        raise NotImplementedError(f"init method has to be implemented for {self}")
```

and probably call `init_weights` inside `from_pretrained` since we do NOT have `FlaxModelMixin__init__`

File: src/diffusers/modeling_flax_utils.py
Comment (patil-suraj) [reply src/diffusers/modeling_flax_utils.py:48]: > but what happens if the model subclass needs to use ignore_for_config for its own variables?
what could be a use-case to ignore the init args for config ?

But also this is a special case for flax so maybe we could hardcode this for now.

File: src/diffusers/modeling_flax_utils.py
Comment (patil-suraj) [inline src/diffusers/modeling_flax_utils.py:63]: (nit) It won't be required here (at least for now), this was added in transformers because of  auto model API

File: src/diffusers/modeling_flax_utils.py
Comment (patil-suraj) [inline src/diffusers/modeling_flax_utils.py:124]: The example should be updated for `diffusers`.

File: src/diffusers/modeling_flax_utils.py
Comment (patil-suraj) [inline src/diffusers/modeling_flax_utils.py:151]: same comment as above

File: src/diffusers/modeling_flax_utils.py
Comment (patil-suraj) [inline src/diffusers/modeling_flax_utils.py:188]: same comment as above.

File: src/diffusers/modeling_flax_utils.py
Comment (patil-suraj) [inline src/diffusers/modeling_flax_utils.py:218]: use `diffusers` checkpoint

File: src/diffusers/modeling_flax_utils.py
Comment (patil-suraj) [inline src/diffusers/modeling_flax_utils.py:250]: This should be updated for `diffusers`. There is no `PreTrainedModel` and `PretrainedConfig`

File: src/diffusers/modeling_flax_utils.py
Comment (patil-suraj) [inline src/diffusers/modeling_flax_utils.py:462]:
```suggestion
        `[`~FlaxModelMixin.from_pretrained`]` class method
```

**Review by patil-suraj — COMMENTED** [review]
Looks very cool in-general, just left some minor nits.

**Comment by patil-suraj** [comment]
> Une Question
In [FlaxModelMixin.from_pretrained](https://github.com/huggingface/diffusers/blob/95073e13b2172bcadddd4ed528360c4f216d753b/src/diffusers/modeling_flax_utils.py#L195), I have NOT added section about checking [missing_keys & mismatched_keys like in trasnformers.flax_from_pretrained](https://github.com/huggingface/transformers/blob/bce36ee065f7749c997be4c30a3f9279df7d5dba/src/transformers/modeling_flax_utils.py#L828-L851).
In short, checking missing_keys & mismatched_keys functionality works by:
random_state = ModelCls.init_weights()
pretrained_state = from_pretrained()
compare keys between random_state VS pretrained_state

La réponse:
Think here we should default to the `_do_init=False` API in `transformers` https://github.com/huggingface/transformers/pull/16148

i.e 
- each model should define `init_weights` method 
- when a model is initialized, weights will never be initialized automatically, the user will always have to call `init_weights`

```python
model = FlaxUnet(.....)
params = model.init_weights()
```

- in `from_pretrained` we should just do `jax.eval_shape` to get the list of param names for the model. This way we can detect missing keys without actually initializing the weights. This helps avoid memory fragmentation.  And then we always force the user to call `init_weights`.

```python
model, params =  FlaxUnet.from_pretrained(.....)
params = model.init_weights(params)
```
Refer to the linked PR to see how it can be implemented.


This way we stay close to JAX/Flax philosophy , nothing is implicitly initialized, the user should be responsible to handle the initialization.

Let me what you think :)

File: src/diffusers/modeling_flax_utils.py
Comment (pcuenca) [reply src/diffusers/modeling_flax_utils.py:48]: Sorry, just to clarify, hardcode how? Using a different list instead of `ignore_for_config` as I was suggesting, or do you have something else in mind? :)

**Comment by pcuenca** [comment]
>  * in `from_pretrained` we should just do `jax.eval_shape` to get the list of param names for the model. This way we can detect missing keys without actually initializing the weights. This helps avoid memory fragmentation.  And then we always force the user to call `init_weights`.
> 
> ```python
> model, params =  FlaxUnet.from_pretrained(.....)
> params = model.init_weights(params)
> ```
> 
> This way we stay close to JAX/Flax philosophy , nothing is implicitly initialized, the user should be responsible to handle the initialization.
> 
> Let me know what you think :)

My initial opinion before I read this comment was that `from_pretrained` should take care of everything so the user does not need to invoke `init_weights`. I think this aligns with the philosophy behind `from_pretrained` in Hugging Face, where it does what it can to get you what you need. In addition, I suppose many users of Flax diffusers will be new to flax. They won't necessarily be familiar with Flax nuances, and probably just want to use a model. I think it's easy to convey that the model parameters are stored separately from the model, but it's harder to explain why they need to follow two steps in order to use them. If we remove one step, it's one step less that could get forgotten or misinterpreted.

There are other libraries that abstract away weight loading, see for example [flaxmodels here](https://github.com/matthias-wright/flaxmodels/blob/main/flaxmodels/vgg/vgg.py#L63-L67).

File: src/diffusers/modeling_flax_utils.py
Comment (mishig25) [reply src/diffusers/modeling_flax_utils.py]: handled in 91559f3107c7d31763094dae4076837d81cbcea1

**Comment by mishig25** [comment]
> i.e 
- each model should define `init_weights` method 
- when a model is initialized, weights will never be initialized automatically, the user will always have to call `init_weights`

```python
model = FlaxUnet(.....)
params = model.init_weights()
```

@patil-suraj, since `FlaxModel` is just `linen.Module`, why [would a user not do](https://flax.readthedocs.io/en/latest/guides/flax_basics.html#model-parameters-initialization):
```
x = random.normal(key1, (...)) # Dummy input
params = FlaxModel.init(key2, x) # Initialization call
```

is it because doing the "normal flax" way (i.e. `FlaxModel.init(key2, x)`) would be bit complicated for the user with all the weights initialization & getting the correct input shape, so we are wrapping it up in a call `FlaxModel.init_weights` ?

File: src/diffusers/modeling_flax_utils.py
Comment (mishig25) [reply src/diffusers/modeling_flax_utils.py:117]: (to keep the github convo readbale, I'm "resolving" the comments about docstrings).
Will work on it once everything else looks good

File: src/diffusers/modeling_flax_utils.py
Comment (mishig25) [reply src/diffusers/modeling_flax_utils.py:181]: (to keep the github convo readbale, I'm "resolving" the comments about docstrings).
Will work on it once everything else looks good

File: src/diffusers/modeling_flax_utils.py
Comment (mishig25) [reply src/diffusers/modeling_flax_utils.py:243]: (to keep the github convo readbale, I'm "resolving" the comments about docstrings).
Will work on it once everything else looks good

File: src/diffusers/modeling_flax_utils.py
Comment (mishig25) [reply src/diffusers/modeling_flax_utils.py:144]: (to keep the github convo readbale, I'm "resolving" the comments about docstrings).
Will work on it once everything else looks good

File: src/diffusers/modeling_flax_utils.py
Comment (mishig25) [reply src/diffusers/modeling_flax_utils.py:211]: (to keep the github convo readbale, I'm "resolving" the comments about docstrings).
Will work on it once everything else looks good

File: src/diffusers/modeling_flax_utils.py
Comment (patrickvonplaten) [inline src/diffusers/modeling_flax_utils.py:34]:
```suggestion
FLAX_WEIGHTS_NAME = "diffusion_flax_model.msgpack"  # TODO should be "diffusion_flax_model.msgpack"
```

File: src/diffusers/modeling_flax_utils.py
Comment (patrickvonplaten) [inline src/diffusers/modeling_flax_utils.py:444]: this looks good to me!

**Comment by patrickvonplaten** [comment]
Super nice PR - will play around with it a bit now!

**Comment by patrickvonplaten** [comment]
@mishig25 for now I'd just add a `init_weights`

> * ld just do `jax.eval_shape` to get the list of param names for the model. This way we can detect missing keys without actually initializing the we

Big +1 here. @mishig25 also happy to add this functionality though in a follow up PR to not blow this one up

**Comment by patrickvonplaten** [comment]
I very much agree with @patil-suraj comments above. Note however that for diffusion models there is no "let's just add a head and fine-tune this" logic => therefore if one does `.from_pretrained(...)` we check automatically that no keys are missing and then there is no need anymore to call `init_weights`. So one doesn't have to call `init_weights` after having run `from_pretrained(...)` usually

**Comment by patrickvonplaten** [comment]
@mishig25 to answer your second question. Indeed we might not even need an `init_weights` function ourselves! We could just use the "native" init function. It's just a bit annoying to have to define the input shape/tensor oneself just for the init. But happy to tackle this in a follow up PR

**Comment by mishig25** [comment]
@patrickvonplaten let's do follow PR to tackle `init_weights` (if needed) so that we can get going with `FlaxModelMixin` from the `main`.

I will fix the docstrings now.
Also, do I need to add soft dependency checks like: `if flax is installed` ?

File: src/diffusers/modeling_flax_utils.py
Comment (patrickvonplaten) [reply src/diffusers/modeling_flax_utils.py:48]: Think "parent", "name" are so Flax specific that we should just hide them in the decorator method

**Comment by patrickvonplaten** [comment]
Works like a charm! You can test it with:

```python
from diffusers.modeling_flax_utils import FlaxModelMixin
from diffusers.configuration_utils import flax_register_to_config, ConfigMixin
import flax.linen as nn
import jax
from jax.random import PRNGKey
import jax.numpy as jnp


@flax_register_to_config
class Adder(nn.Module, FlaxModelMixin, ConfigMixin):
    vocab_size: int
    hidden_size: int = 8
    initializer_range: float = 0.4

    def setup(self):
        self.word_embeddings = nn.Embed(
            self.config.vocab_size,
            self.config.hidden_size,
            embedding_init=jax.nn.initializers.normal(stddev=self.config.initializer_range),
        )

    def __call__(self, input_tensor):
        return self.word_embeddings(input_tensor) + 6


adder = Adder(5, hidden_size=10)
ones = jnp.ones((1, 1), dtype="i4")
rngs = {"params": PRNGKey(0)}
params = adder.init(rngs, ones)

adder.save_pretrained("dummy", params)
adder_new, params = Adder.from_pretrained("./dummy")
```

**Comment by patrickvonplaten** [comment]
## Update

I've removed the `ignore_for_config` for now as we've used this mostly for the schedulers `__init__` method. I don't think Flax models will need this soon. Also as @patil-suraj, `("parent", "name")` are now hard-coded just as it's done in Flax' original codebase.

Some tweaks to the constructor were necessary, but apart from this it looks great! Awesome job @mishig25

**Review by patrickvonplaten — APPROVED** [review]
Once the weight name is changed to `"diffusion_flax_model.msgpack"` let's merge this PR to unblock a couple of other PRs :-) 

## TODOs 

- [ ] Add functionality that automatically detects missing keys and in this case, I think we should just throw an error in `from_pretrained(...)` for now. This way we can follow @pcuenca suggestion and don't have to manually run `init(...)` after having run `from_pretrained(...)`
- [ ] Tests - we should add some tests here

**Comment by patrickvonplaten** [comment]
@pcuenca could you do one more review here and then we could merge I think :-)

**Review by pcuenca — APPROVED** [review]
Works great for me! (Testing with `FlaxUNet2DConditionModel` in another branch)

Just one thing: I now get this warning when loading the pretrained model:
```
{'name', 'parent'} was not found in config. Values will be initialized to default values.
```

Could we maybe exclude `"parent"` and `"name"` here https://github.com/huggingface/diffusers/blob/25ab3cad11e624fd16d16283fa7224fe082881b8/src/diffusers/configuration_utils.py#L271
 (However I see that `extract_init_dict` is not Flax specific, so not sure about how to deal with it)

**Comment by patrickvonplaten** [comment]
> ow get this warning when loading the pretraine

Hmm double checking maybe I added a bug somewhere - this warning shouldn't happen. Thanks for the heads-up

**Comment by pcuenca** [comment]
> Hmm double checking maybe I added a bug somewhere - this warning shouldn't happen. Thanks for the heads-up

The removal of `ignore_for_config` maybe? They are not excluded from the expected list any more.

**Comment by patrickvonplaten** [comment]
@pcuenca - actually reverting some logic a bit. Will add a `_flax_internal_args` class attribute since now we need the name in two places

File: src/diffusers/modeling_flax_utils.py
Comment (patrickvonplaten) [inline src/diffusers/modeling_flax_utils.py:48]: @pcuenca - would that be ok for you?

**Comment by mishig25** [comment]
@patil-suraj docstrings are updated.

Please let me know if/when I should merge

File: src/diffusers/modeling_flax_utils.py
Comment (pcuenca) [reply src/diffusers/modeling_flax_utils.py:48]: Yes, that feels much better than using something that subclasses could change. I think it's perfect! :)

**Comment by kashif** [comment]
@mishig25 can you add the imports to init file and create dummy imports?

File: src/diffusers/modeling_flax_utils.py
Comment (pcuenca) [inline src/diffusers/modeling_flax_utils.py:213]: Does this work? Just curious, I don't know if it does.

File: src/diffusers/modeling_flax_utils.py
Comment (mishig25) [reply src/diffusers/modeling_flax_utils.py:102]: sounds good, lets update the docstrings when "CompVis/stable-diffusion-v1-4" gets updated

File: src/diffusers/modeling_flax_utils.py
Comment (mishig25) [reply src/diffusers/modeling_flax_utils.py]: Removed it https://github.com/huggingface/diffusers/pull/493/commits/82311a5dc3fb4a310adc70d0192588feea849ef9 because we don't currently have conversion script between flax <> pt

**Comment by mishig25** [comment]
> @mishig25 can you add the imports to init file and create dummy imports?

Added FlaxModelMixin to __init__.py a94968f434022298a9f4a2e40409ea0465561947
However, I am not sure where we would add flax dummy object?

**Comment by kashif** [comment]
@mishig25 run: `mmake fix-copies`

**Comment by mishig25** [comment]
@kashif nothing happened
```
(base) mishig@vorace:~/diffusers$ make fix-copies
python utils/check_dummies.py --fix_and_overwrite
(base) mishig@vorace:~/diffusers$ 
```

**Comment by kashif** [comment]
@mishig25 perhaps first update your branch and then i believe you need to put the imports inside the `is_flax_available...` stuff... have a look at another init file

**Comment by mishig25** [comment]
@kashif thanks a lot for the pointers! I've done so after rebasing from `main`

File: src/diffusers/__init__.py
Comment (patrickvonplaten) [inline src/diffusers/__init__.py:66]: Perfect!

File: src/diffusers/modeling_flax_utils.py
Comment (patrickvonplaten) [reply src/diffusers/modeling_flax_utils.py:102]: Good point @pcuenca! We can also work with `allow/reject` regexes so that PyTorch wouldn't download Flax weights - this way everything could be in the same branch. Note that this is only relevant for the pipeline yet though

**Comment by patrickvonplaten** [comment]
Once tests are green - I think we can merge :partying_face:

**Pull request merged**
